x/wasm/keeper/gas_register.go

```diff
@@ -0,0 +1,128 @@
+	// Rough timing have 88k gas at 90us, which is equal to 1k sdk gas... (one read)
+	//
+	// Please note that all gas prices returned to the wasmer engine should have this multiplied
+	DefaultGasMultiplier uint64 = 100
+	// DefaultInstanceCost is how much SDK gas we charge each time we load a WASM instance.
+	// Creating a new instance is costly, and this helps put a recursion limit to contracts calling contracts.
+	DefaultInstanceCost uint64 = 40_000
+	// DefaultCompileCost is how much SDK gas we charge *per byte* for compiling WASM code.
+	DefaultCompileCost uint64 = 2
+	// DefaultEventAttributeDataCost is how much SDK gas we charge *per byte* for attribute data in events.
+	// This is len(key) + len(value)
+	DefaultEventAttributeDataCost uint64 = 1
+	// DefaultPerAttributeCost is how much SDK gas we charge per attribute count.
+	DefaultPerAttributeCost uint64 = 10
+	// DefaultEventAttributeDataFreeTier number of bytes of attribute data we do not charge.
+	DefaultEventAttributeDataFreeTier = 100
+)
+
+type GasRegister struct {
+	instanceCost  sdk.Gas
+	compileCost   sdk.Gas
+	gasMultiplier sdk.Gas
+
+	eventPerAttributeCost      sdk.Gas
+	eventAttributeDataCost     sdk.Gas
+	eventAttributeDataFreeTier int
+}
+
+func DefaultGasRegister() GasRegister {
+	return GasRegister{
+		instanceCost:               DefaultInstanceCost,
+		compileCost:                DefaultCompileCost,
+		gasMultiplier:              DefaultGasMultiplier,
+		eventPerAttributeCost:      DefaultPerAttributeCost,
+		eventAttributeDataCost:     DefaultEventAttributeDataCost,
+		eventAttributeDataFreeTier: DefaultEventAttributeDataFreeTier,
+	}
+}
+func NewGasRegister(
+	instanceCost sdk.Gas,
+	compileCost sdk.Gas,
+	gasMultiplier sdk.Gas,
+	eventAttributeCountCost sdk.Gas,
+	eventAttributeLengthCost sdk.Gas,
+	freeTierAttributeData int,
+) GasRegister {
+	return GasRegister{
+		instanceCost:               instanceCost,
+		compileCost:                compileCost,
+		gasMultiplier:              gasMultiplier,
+		eventPerAttributeCost:      eventAttributeCountCost,
+		eventAttributeDataCost:     eventAttributeLengthCost,
+		eventAttributeDataFreeTier: freeTierAttributeData,
+	}
+}
+
+func (g GasRegister) NewContractInstanceCost(pinned bool, msgLen int, labelLength int) storetypes.Gas {
+	return g.InstantiateContractCost(pinned, msgLen)
+}
+
+func (g GasRegister) CompileCost(byteLength int, sourceCodeUrlLen int, builderLen int) storetypes.Gas {
+	return g.compileCost * uint64(byteLength)
+}
+
+func (g GasRegister) InstantiateContractCost(pinned bool, msgLen int) sdk.Gas {
+	if pinned {
+		return 0
+	}
+	return g.instanceCost
+}
+
+func (g GasRegister) ReplyCost(pinned bool, reply wasmvmtypes.Reply) sdk.Gas {
+	var eventGas sdk.Gas
+	msgLen := len(reply.Result.Err)
+	if reply.Result.Ok != nil {
+		msgLen += len(reply.Result.Ok.Data)
+		for _, e := range reply.Result.Ok.Events {
+			msgLen += len(e.Type)
+			eventGas += g.EventCosts(e.Attributes)
```

Comment:
charging each event separately (not combined) will lead to some skew.
The "free tier" is no longer calculated on the total data returned but on each event.

So, I could have 10 events with 100 bytes of attributes and stay in the free tier, which is not desired. I would combine all attributes of all events and then calculate `EventsCost`.